File: Lab 4 (13 February 2025)/iterative quick sort.cpp
#include <iostream>
#include <stack>
using namespace std;

int partition(int arr[], int low, int high)
{
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j <= high - 1; j++)
    {
        if (arr[j] < pivot)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSortIterative(int arr[], int l, int h)
{
    stack<int> s;

    s.push(l);
    s.push(h);
    while (!s.empty())
    {
        h = s.top(); s.pop();
        l = s.top(); s.pop();
        int p = partition(arr, l, h);
        if (p - 1 > l)
        {
            s.push(l);
            s.push(p - 1);
        }
        if (p + 1 < h)
        {
            s.push(p + 1);
            s.push(h);
        }
    }
}

int main()
{
    int arr[] = {4, 2, 6, 9, 1, 3, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    quickSortIterative(arr, 0, n - 1);
    cout << "Sorted array: ";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
    return 0;
}

File: Lab 4 (13 February 2025)/iterative quick sort.py
import time
import random
import matplotlib.pyplot as p

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return i + 1

def iterative_quick_sort(arr):
    stack = [(0, len(arr) - 1)]
    
    while stack:
        low, high = stack.pop()
        if low < high:
            pi = partition(arr, low, high)
            stack.append((low, pi - 1))
            stack.append((pi + 1, high))

sizes = list(range(1000, 11000, 1000))
times = []

for size in sizes:
    arr = [random.randint(1, 100000) for _ in range(size)]
    start = time.time()
    iterative_quick_sort(arr)
    end = time.time()
    times.append(end - start)

p.plot(sizes, times, marker='o', label='Iterative Quick Sort (Avg: O(n log n))')
p.xlabel('Array Size (n)')
p.ylabel('Execution Time (seconds)')
p.title('Time Complexity of Iterative Quick Sort')
p.grid(True)
p.legend()
p.show()